### PR TITLE
Refactor JSON schema validation

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This PR moves around some implementation-details for creating EntityTypes.

Before this PR, the `EntityType` JSON Schema was validated in the database client, creating a `jsonSchemaCompiler` dependency.
In this PR, the JSON Schema validation happens from the `EntityType` model class, so that the database client doesn't need to use the `jsonSchemaCompiler`.

These changes also prompted a couple changes to the inheritance field resolver (stylistic choice).

We will now have a more centralized location to do JSON schema validation as well.

## 🐾 Next steps

- To implement better inheritance chain resolving see [asana task](https://app.asana.com/0/1201095311341924/1201715431493439/f) (_internal_)
- We could consolidate more of the Update/create logic in the postgres client for EntityType, I decided not to do this here.

## 🔍 What does this change?

- Move JSON schema validation up from the database client to the EntityType model class
- Remove helper function that did not belong next to the SQL queries
- Spread EntityType inheritance field resolvers in the GQL resolvers object (better ergonomics, thanks for showing me, Ben!)

### 📜 Does this require a change to the docs?

- The docs for x need updating to explain that y

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1201849001249402/f) (_internal_)

## 🛡 What tests cover this?

- Existing integration tests

## ❓ How to test this?

This doesn't change any functionality

### CODE REVIEW

File: packages/hash/api/src/model/entityType.model.ts
Comment (benwerner01) [inline packages/hash/api/src/model/entityType.model.ts:145]: Might be out of scope of this PR, but I don't think this should be a static method. If you agree it'd be nice to change that or address it as part of a follow-up task.

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (benwerner01) [inline packages/hash/api/src/graphql/resolvers/index.ts:212]: Nice 😉

File: packages/hash/api/src/db/postgres/client.ts
Comment (benwerner01) [inline packages/hash/api/src/db/postgres/client.ts:109]:
```suggestion
      schema.$id = generateSchema$id(accountId, entityTypeId);
```

File: packages/hash/api/src/db/postgres/client.ts
Comment (benwerner01) [inline packages/hash/api/src/db/postgres/client.ts:361]:
```suggestion
    schema.$id = generateSchema$id(entity.accountId, entityId);
```

File: packages/hash/api/src/lib/schemas/jsonSchema.ts
Comment (benwerner01) [inline packages/hash/api/src/lib/schemas/jsonSchema.ts:200]: Can we make it an optional argument? I realise we won't always be able to know the `accountId` and `entityId` when this is called, but when we do it'd be nice to just be able to pass the `$id` directly.

**Review by benwerner01 — APPROVED** [review]
Nice work Ahmad, this looks good to me 👍

Left a couple of suggestions for further improvement, some might be beyond the scope of this PR so we could address it in follow-up. Thanks :)

File: packages/hash/api/src/model/entityType.model.ts
Comment (thehabbos007) [reply packages/hash/api/src/model/entityType.model.ts:145]: I agree with this, I have a question regarding the resolver:
Since we would have to fetch the `EntityType` if we make this change, should we then only allow updates if the user's `accountId` is equal to that of the `EntityType`? 
If I am not mistaken, it is currently possible to change `EntityType`s that the account did not create. What are your thoughts on this?

File: packages/hash/api/src/lib/schemas/jsonSchema.ts
Comment (thehabbos007) [reply packages/hash/api/src/lib/schemas/jsonSchema.ts:200]: Good idea. Should we still make the title the default value in case `$id` is not set?

File: packages/hash/api/src/model/entityType.model.ts
Comment (benwerner01) [reply packages/hash/api/src/model/entityType.model.ts:145]: My instinct is this is not to implement such a check now, as it'll be addressed when we implement the permissions model. That's when we should discuss where the best places would be to perform permission checks.

Right now you're right, a user can in theory update entities/links/entityTypes that aren't in an account they have access to. There's a basic check in the GQL queries that tries to prevent the user from doing things in other accounts, but there's likely edge cases this middleware isn't accounting for.

File: packages/hash/api/src/lib/schemas/jsonSchema.ts
Comment (benwerner01) [reply packages/hash/api/src/lib/schemas/jsonSchema.ts:200]: If `$id` is required by the scheme validator, I see no other way - yeah

Comment (thehabbos007) [reply packages/hash/api/src/lib/schemas/jsonSchema.ts:200]: Seems that it is actually optional. I'll add it as an optional argument!

Comment (benwerner01) [reply packages/hash/api/src/lib/schemas/jsonSchema.ts:200]: Ah even better, in that case I think we can avoid setting `$id` to `title` entirely 👍

File: packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts
Comment (benwerner01) [inline packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts:37]: Is there a reasoning why only the update occurs in the transaction? Is it to reduce the size the transaction?

Comment (benwerner01) [inline packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts:37]: I think the following would be more readable:
```suggestion
  return await db.transaction(async (conn) => {
    await entityType.update(conn, {
      accountId,
      updatedByAccountId: user.accountId,
      createdByAccountId: user.entityId,
      schema,
    });

    return entityType.toGQLEntityType();
  });
```

File: packages/hash/api/src/lib/schemas/jsonSchema.ts
Comment (benwerner01) [inline packages/hash/api/src/lib/schemas/jsonSchema.ts:179]: Why accept `null` as a value for `$id`?
```suggestion
    $id?: string;
```
(This would also save you from having to do `$id: $id ?? undefined` later on)

File: packages/hash/api/src/model/entityType.model.ts
Comment (benwerner01) [inline packages/hash/api/src/model/entityType.model.ts:168]: **Blocking:** Because this is now no longer a static function, we need to update the member variables that are affected by the db update. We could either do this individually by setting things like `properties`, `updatedAt`, etc - or we could make use of lodash `merge`:

```suggestion
    const updatedDbEntityType = await client.updateEntityType({
      ...params,
      entityId: this.entityId,
      schema,
    });
    
    merge(this, new EntityType(updatedDbEntityType));
    
    return this;
```

This is what we are also currently doing in `Entity.updateProperties`.

**Review by benwerner01 — CHANGES_REQUESTED** [review]
Thanks Ahmad for addressing my suggestions :)

There's one thing that I think should be addressed regarding the new non-static `update` method of the `EntityType` model class that has was introduced as part of that.

But once that's sorted, this PR is good to merge 👍

File: packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts:37]: No reason for this. I'll wrap it all in the transaction :)

Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts:37]: (I now realise you were doing this to make use of the return type of the `update` method, but once [this](https://github.com/hashintel/hash/pull/369#discussion_r820528519) has been addressed you'll no longer have to do that)

Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts:37]: Okay, well I think it's equivalent so doesn't require changing - it's just not consistent with other resolvers where we do everything in one transaction (including the fetching).

Once we consistently lock entities that we are updating on retrieval, the fetch will need to occur in the transaction though.

Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts:37]: Addressed along with having transaction around all of resolver :+1:

File: packages/hash/api/src/lib/schemas/jsonSchema.ts
Comment (thehabbos007) [reply packages/hash/api/src/lib/schemas/jsonSchema.ts:179]: Good point, it isn't a function param, so it can just be left out instead of set to null :+1:

File: packages/hash/api/src/model/entityType.model.ts
Comment (thehabbos007) [reply packages/hash/api/src/model/entityType.model.ts:168]: Of course, sorry for missing that! Thank you :)

Comment (thehabbos007) [reply packages/hash/api/src/model/entityType.model.ts:168]: I think my instinct was to not mutate the calling object, but since we do this throughout the model classes, it will be changed here :)

Comment (benwerner01) [reply packages/hash/api/src/model/entityType.model.ts:169]: What would be the reasoning for not mutating the calling object? I feel if we don't mutate it we risk that the caller of the model class instance makes use of outdated data (for instance the incorrect `updatedAt`, the outdated `schema`, etc).

I appreciate that this wouldn't be an issue if we were to consistently use the return value of an update function, but this will be hard to enforce and could therefore quickly lead to bugs.

Comment (thehabbos007) [reply packages/hash/api/src/model/entityType.model.ts:169]: Yes, it makes sense for our model classes we're doing it this way. My instinct stems from other projects where it lead to some unexpected behaviour, so it was more general instinct - not bound to our context.

I don't think those worries exist for our model classes since they are short lived and owned by a single resolver anyway :)

File: packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts
Comment (benwerner01) [inline packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts:18]: Nit:
```suggestion
    const entityType = await EntityType.getEntityType(conn, {
```

**Review by benwerner01 — APPROVED** [review]
Thanks Ahmad, spotted one last nit but nothing that blocks merging 👍

File: packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/entityType/updateEntityType.ts:18]: Let me get this in as well :)

**Review by benwerner01 — APPROVED** [review]
🎉

**Pull request merged**
